packages/foundry/contracts/FYToken.sol

```diff
@@ -129,6 +191,35 @@ contract FYToken is IFYToken, IERC3156FlashLender, AccessControl, ERC20Permit, C
         emit Redeemed(msg.sender, to, amount_, redeemed);
     }
 
+    ///@dev returns the maximum withdrawable amount for the address holder in terms of the underlying
+    function maxWithdraw(address holder) external override returns (uint256 maxUnderlyingAmount) {
+        return _convertToUnderlying(_balanceOf[holder]);
+    }
+
+    ///@dev returns the amount of the principal withdrawable in terms of the underlying
+    function previewWithdraw(uint256 underlyingAmount) external override beforeMaturity returns (uint256 principalAmount) {
+        return _previewWithdraw(underlyingAmount);
+    }
+
+    ///@dev returns the amount of underlying redeemable in terms of the principal
+    function _previewWithdraw(uint256 underlyingAmount) internal returns (uint256 principalAmount) {
+        principalAmount = underlyingAmount.wdiv(_accrual());
+    }
+
+    function withdraw(uint256 underlyingAmount, address receiver, address holder) external override afterMaturity returns (uint256 principalAmount) {
+        if (msg.sender != holder) {
+            uint256 allowed = _allowance[holder][msg.sender];
+
+            if (allowed != type(uint256).max) _allowance[holder][msg.sender] = allowed - principalAmount;
```

Comment:
`principalAmount` will always be 0 here.

if `underlyingAmount` is 0, then we skip the allowance check and `principalAmount = _balanceOf[address(this)]`
else, we `principalAmount = _convertToPrincipal(underlyingAmount)` and then check for allowance of `principalAmount`

Could you please code the unit tests that catch this bug first, and then fix it?

Reply:
I think I found another problem. If the allowance is subtracted before the call to burn, and say hypothetically the amount provided is the same as the allowance, then burn will revert on insufficient allowance since it checks it there.

Reply:
Since `_burn` checks the allowance and resets it internally, I think we can avoid this check.

Reply:
Good point